### PR TITLE
[BACKLOG-4526] - Dev subtask for Create public REST interfaces for Us…

### PR DESCRIPTION
**Pull request by apratkin**
…er Create

### CODE REVIEW

**Comment by apratkin** [comment]
@brendan-lynch could you review?

File: extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java [hunk 1]
```diff
@@ -485,6 +486,34 @@ public Response createUser( @QueryParam ( "tenant" ) String tenantPath, User use
   }
 
   /**
+   * Create new user with specified name and password
```
Comment (brendan-lynch) [inline extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java]: Comments need to include the response and request examples for using this call

File: extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java [hunk 2]
```diff
@@ -485,6 +486,34 @@ public Response createUser( @QueryParam ( "tenant" ) String tenantPath, User use
...
+  @POST
+  @Path( "/user" )
+  @Consumes( { WILDCARD } )
+  @StatusCodes( { 
+    @ResponseCode( code = 200, condition = "Successfully created new user." ), 
+    @ResponseCode( code = 403, condition = "A non administrative user is trying to access this endpoint." ), 
+    @ResponseCode( code = 500, condition = "Unable to create User objects." ) 
+  } )
+  public Response createUserPublicApi( User user ) {
+    if ( canAdminister() ) {
```
Comment (brendan-lynch) [inline extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java]: canAdminister logic should exist in the service layer. The service layer can throw a SecurityException in the canAdminister else block, which the catch clause in the resource maps to the web application exception 403 error code.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 3]
```diff
@@ -17,15 +17,46 @@
...
+    return nameValid && passValid;
+  }
+
+  public void createUser( User user ) throws UnauthorizedException, Exception {
+    if ( userValid( user ) ) {
+      IUserRoleDao roleDao =
+          PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
+      roleDao.createUser( null, user.getUserName(), user.getPassword(), "", new String[0] );
+    } else {
+      throw new ValidationFailedException();
```
Comment (brendan-lynch) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: If we are going to throw a validation exception failure here we should also catch this in the resource layer and map it to a response code, to let the user know that the username contains invalid characters. I like this exception for validation though.

**Comment by apratkin** [comment]
Tested:
1.	Try to create user - ok
2.	Try to create user with name which already exist - 500 response code
3.	Try to create user not by administrator - 403
4.	Try to create user with name on Russian - ok
5.	Try to create user without password (empty) – 400
6.	Try to create user without password (null) – 400
7.	Try to create user with empty name (empty) – 400
8.	Try to create user with empty name (null) – 400
9.	Try to create user with name more than 300 symbols – ok
We do not have limitation on length of name
10.	Try to create user with reserved symbols (/,\,\t,\r,\n) at name– 400

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java [hunk 4]
```diff
@@ -360,4 +362,56 @@ public void testSetLogicalRolesSecurityException() {
...
+  public void testCreateUserValidationFailed() throws Exception {
+    UserRoleDaoService mockService = mock( UserRoleDaoService.class );
+    doThrow( new UserRoleDaoService.ValidationFailedException() ).when( mockService ).createUser( any( User.class ) );
+    UserRoleDaoResource resource =
+        new UserRoleDaoResource( roleBindingDao, tenantManager, systemRoles, adminRole, mockService );
+    try {
+      resource.createUserPublicApi( new User( "\\/validation", "failed" ) );
+    } catch ( WebApplicationException e ) {
+      assertEquals( Response.Status.BAD_REQUEST.getStatusCode(), e.getResponse().getStatus() );
+      return;
```
Comment (brendan-lynch) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java]: remove the return statement and the assert failed

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java [hunk 5]
```diff
@@ -360,4 +362,56 @@ public void testSetLogicalRolesSecurityException() {
...
+    doThrow( new SecurityException() ).when( mockService ).createUser( any( User.class ) );
+    UserRoleDaoResource resource =
+        new UserRoleDaoResource( roleBindingDao, tenantManager, systemRoles, adminRole, mockService );
+    try {
+      resource.createUserPublicApi( new User( "not", "admin" ) );
+    } catch ( WebApplicationException e ) {
+      assertEquals( Response.Status.FORBIDDEN.getStatusCode(), e.getResponse().getStatus() );
+      return;
+    }
+    Assert.fail();
```
Comment (brendan-lynch) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java]: remove the return statement and the assert failed

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java [hunk 6]
```diff
@@ -360,4 +362,56 @@ public void testSetLogicalRolesSecurityException() {
...
+    UserRoleDaoService mockService = mock( UserRoleDaoService.class );
+    doThrow( new AlreadyExistsException( "message" ) ).when( mockService ).createUser( any( User.class ) );
+    UserRoleDaoResource resource =
+        new UserRoleDaoResource( roleBindingDao, tenantManager, systemRoles, adminRole, mockService );
+    try {
+      resource.createUserPublicApi( new User( "user", "duplicate" ) );
+    } catch ( WebApplicationException e ) {
+      assertEquals( Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), e.getResponse().getStatus() );
+      return;
+    }
```
Comment (brendan-lynch) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java]: Same comment as above

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoServiceTest.java [hunk 7]
```diff
@@ -523,4 +526,39 @@ public void testGetRoleBindingStructSecurityException() {
...
+    return false;
+  }
+
+  @Test
+  public void testCreateUserValidation() {
+    List<String> reservedChars = Arrays.asList( "/,\\,\t,\r,\n".split( "," ) );
+    IUnifiedRepository repo = mock( IUnifiedRepository.class );
+    doReturn( reservedChars ).when( repo ).getReservedChars();
+    PentahoSystem.registerObject( repo );
+    UserRoleDaoService service = new UserRoleDaoService(){
```
Comment (brendan-lynch) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoServiceTest.java]: See my unit test examples, we should not be mocking this, the code in private methods should be exercised through the call to the method that accesses them. Mock the objects that PentahoSystem.get() returns.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 8]
```diff
@@ -170,7 +202,7 @@ public void setLogicalRoles( LogicalRoleAssignments roleAssignments ) throws Sec
     }
   }
 
-  private boolean canAdminister() {
+  protected boolean canAdminister() {
```
Comment (brendan-lynch) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: This should continue to be a private method, there is no business exposing this to the package just for unit testing when we have viable alternatives using pentaho system.

**Pull request merged**